packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -140,18 +161,117 @@ export class TreeViewConfiguration<TSchema extends ImplicitFieldSchema = Implici
 		this.schema = config.schema;
 		this.enableSchemaValidation = config.enableSchemaValidation;
 
-		// Ensure all reachable schema are marked as most derived.
-		// This ensures if multiple schema extending the same schema factory generated class are present (or have been constructed, or get constructed in the future),
-		// an error is reported.
-		walkFieldSchema(config.schema, markSchemaMostDerived);
+		// Ambiguity errors are lower priority to report than invalid schema errors, so collect these in an array and report them all at once.
+		const ambiguityErrors: string[] = [];
+
+		walkFieldSchema(config.schema, {
+			// Ensure all reachable schema are marked as most derived.
+			// This ensures if multiple schema extending the same schema factory generated class are present (or have been constructed, or get constructed in the future),
+			// an error is reported.
+
+			node: markSchemaMostDerived,
+			allowedTypes(types): void {
+				if (config.ensureUnambiguous) {
+					checkUnion(types, ambiguityErrors);
+				}
+			},
+		});
+
+		if (ambiguityErrors.length !== 0) {
+			throw new UsageError(`Ambagious Schema Found:\n${ambiguityErrors.join("\n")}`);
+		}
+
 		// Eagerly perform this conversion to surface errors sooner.
 		toFlexSchema(config.schema);
 	}
 }
 
+function checkUnion(union: Iterable<TreeNodeSchema>, errors: string[]): void {
+	const checked: Set<TreeNodeSchema> = new Set();
+	const maps: TreeNodeSchema[] = [];
+	const arrays: TreeNodeSchema[] = [];
+
+	const objects: ObjectNodeSchema[] = [];
+	// Map from key to schema using that key
+	const allObjectKeys: Map<string, ObjectNodeSchema[]> = new Map();
+
+	for (const schema of union) {
+		if (checked.has(schema)) {
+			throw new UsageError(`Duplicate schema in allowed types: ${schema.identifier}`);
+		}
+		checked.add(schema);
+
+		if (union instanceof LeafNodeSchema) {
+			// nothing to do
+		} else if (isObjectNodeSchema(schema)) {
+			objects.push(schema);
+			for (const key of schema.fields.keys()) {
+				getOrAddEmptyToMap(allObjectKeys, key).push(schema);
+			}
+		} else if (schema.kind === NodeKind.Array) {
+			arrays.push(schema);
+		} else {
+			assert(schema.kind === NodeKind.Map, "invalid schema");
+			maps.push(schema);
+		}
+	}
+
+	if (arrays.length > 1) {
+		errors.push(
+			`More than one kind of array allowed within union (${JSON.stringify(arrays.map((s) => s.identifier))}). This would require type disambiguation which is not supported by arrays during import or export to JSON.`,
```

Comment:
This is a case where I wish we tracked issues in GitHub. Then we could link to the work tracking type disambiguation. Oh well :(

Reply:
I don't expect this case to ever be removed. When an array is empty `[]` does not provide any possible way to differentiate between an array of Foo or an array of Bar.

Reply:
Oops, left this in the wrong place. We had some errors with notes about potential future improvements. I was referring to those.